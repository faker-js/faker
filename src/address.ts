import type { Faker } from '.';
import type { Fake } from './fake';
import type { Helpers } from './helpers';

let f: Fake['fake'];

export class Address {
  readonly Helpers: Helpers;

  constructor(private readonly faker: Faker) {
    f = this.faker.fake;
    this.Helpers = this.faker.helpers;

    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Address.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }

    // TODO @Shinigami92 2022-01-13: Need better strategy
    // @ts-expect-error
    this.direction.schema = {
      description:
        'Generates a direction. Use optional useAbbr bool to return abbreviation',
      sampleResults: ['Northwest', 'South', 'SW', 'E'],
    };
    // @ts-expect-error
    this.cardinalDirection.schema = {
      description:
        'Generates a cardinal direction. Use optional useAbbr boolean to return abbreviation',
      sampleResults: ['North', 'South', 'E', 'W'],
    };
    // @ts-expect-error
    this.ordinalDirection.schema = {
      description:
        'Generates an ordinal direction. Use optional useAbbr boolean to return abbreviation',
      sampleResults: ['Northwest', 'Southeast', 'SW', 'NE'],
    };
  }

  /**
   * Generates random zipcode from format. If format is not specified, the
   * locale's zip format is used.
   *
   * @method faker.address.zipCode
   * @param format
   */
  zipCode(format?: string) {
    // if zip format is not specified, use the zip format defined for the locale
    if (typeof format === 'undefined') {
      const localeFormat = this.faker.definitions.address.postcode;
      if (typeof localeFormat === 'string') {
        format = localeFormat;
      } else {
        format = this.faker.random.arrayElement(localeFormat);
      }
    }
    return this.Helpers.replaceSymbols(format);
  }

  /**
   * Generates random zipcode from state abbreviation. If state abbreviation is
   * not specified, a random zip code is generated according to the locale's zip format.
   * Only works for locales with postcode_by_state definition. If a locale does not
   * have a postcode_by_state definition, a random zip code is generated according
   * to the locale's zip format.
   *
   * @method faker.address.zipCodeByState
   * @param state
   */
  zipCodeByState(state: string) {
    const zipRange = this.faker.definitions.address.postcode_by_state[state];
    if (zipRange) {
      return this.faker.datatype.number(zipRange);
    }
    return this.faker.address.zipCode();
  }

  /**
   * Generates a random localized city name. The format string can contain any
   * method provided by faker wrapped in `{{}}`, e.g. `{{name.firstName}}` in
   * order to build the city name.
   *
   * If no format string is provided one of the following is randomly used:
   *
   * * `{{address.cityPrefix}} {{name.firstName}}{{address.citySuffix}}`
   * * `{{address.cityPrefix}} {{name.firstName}}`
   * * `{{name.firstName}}{{address.citySuffix}}`
   * * `{{name.lastName}}{{address.citySuffix}}`
   * * `{{address.cityName}}` when city name is available
   *
   * @method faker.address.city
   * @param format
   */
  city(format?: string | number) {
    const formats = [
      '{{address.cityPrefix}} {{name.firstName}}{{address.citySuffix}}',
      '{{address.cityPrefix}} {{name.firstName}}',
      '{{name.firstName}}{{address.citySuffix}}',
      '{{name.lastName}}{{address.citySuffix}}',
    ];

    if (!format && this.faker.definitions.address.city_name) {
      formats.push('{{address.cityName}}');
    }

    if (typeof format !== 'number') {
      format = this.faker.datatype.number(formats.length - 1);
    }

    return f(formats[format]);
  }

  /**
   * Return a random localized city prefix
   *
   * @method faker.address.cityPrefix
   */
  cityPrefix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_prefix
    );
  }

  /**
   * Return a random localized city suffix
   *
   * @method faker.address.citySuffix
   */
  citySuffix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_suffix
    );
  }

  /**
   * Returns a random city name
   *
   * @method faker.address.cityName
   */
  cityName(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_name
    );
  }

  /**
   * Returns a random localized street name
   *
   * @method faker.address.streetName
   */
  streetName(): string {
    let result: string;
    let suffix = this.faker.address.streetSuffix();
    if (suffix !== '') {
      suffix = ' ' + suffix;
    }

    switch (this.faker.datatype.number(1)) {
      case 0:
        result = this.faker.name.lastName() + suffix;
        break;
      case 1:
        result = this.faker.name.firstName() + suffix;
        break;
    }
    return result;
  }

  //
  // TODO: change all these methods that accept a boolean to instead accept an options hash.
  //

  /**
   * Returns a random localized street address
   *
   * @method faker.address.streetAddress
   * @param useFullAddress
   */
  streetAddress(useFullAddress: boolean = false): string {
    let address = '';
    switch (this.faker.datatype.number(2)) {
      case 0:
        address =
          this.Helpers.replaceSymbolWithNumber('#####') +
          ' ' +
          this.faker.address.streetName();
        break;
      case 1:
        address =
          this.Helpers.replaceSymbolWithNumber('####') +
          ' ' +
          this.faker.address.streetName();
        break;
      case 2:
        address =
          this.Helpers.replaceSymbolWithNumber('###') +
          ' ' +
          this.faker.address.streetName();
        break;
    }
    return useFullAddress
      ? address + ' ' + this.faker.address.secondaryAddress()
      : address;
  }

  /**
   * streetSuffix
   *
   * @method faker.address.streetSuffix
   */
  streetSuffix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.street_suffix
    );
  }

  /**
   * streetPrefix
   *
   * @method faker.address.streetPrefix
   */
  streetPrefix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.street_prefix
    );
  }

  /**
   * secondaryAddress
   *
   * @method faker.address.secondaryAddress
   */
  secondaryAddress() {
    return this.Helpers.replaceSymbolWithNumber(
      this.faker.random.arrayElement(['Apt. ###', 'Suite ###'])
    );
  }

  /**
   * county
   *
   * @method faker.address.county
   */
  county(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.county
    );
  }

  /**
   * country
   *
   * @method faker.address.country
   */
  country(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.country
    );
  }

  /**
   * countryCode
   *
   * @method faker.address.countryCode
   * @param alphaCode default alpha-2
   */
  countryCode(alphaCode: string = 'alpha-2'): string {
    if (alphaCode === 'alpha-2') {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.country_code
      );
    }

    if (alphaCode === 'alpha-3') {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.country_code_alpha_3
      );
    }

    return this.faker.random.arrayElement(
      this.faker.definitions.address.country_code
    );
  }

  /**
   * state
   *
   * @method faker.address.state
   * @param useAbbr
   */
  // TODO christopher 2022-01-13: useAbbr not in use
  state(useAbbr?: boolean): string {
    return this.faker.random.arrayElement(this.faker.definitions.address.state);
  }

  /**
   * stateAbbr
   *
   * @method faker.address.stateAbbr
   */
  stateAbbr(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.state_abbr
    );
  }

  /**
   * latitude
   *
   * @method faker.address.latitude
   * @param max default is 90
   * @param min default is -90
   * @param precision default is 4
   */
  latitude(max: number = 90, min: number = -90, precision: number = 4): string {
    return this.faker.datatype
      .number({
        max: max,
        min: min,
        precision: parseFloat((0.0).toPrecision(precision) + '1'),
      })
      .toFixed(precision);
  }

  /**
   * longitude
   *
   * @method faker.address.longitude
   * @param max default is 180
   * @param min default is -180
   * @param precision default is 4
   */
  longitude(
    max: number = 180,
    min: number = -180,
    precision: number = 4
  ): string {
    return this.faker.datatype
      .number({
        max: max,
        min: min,
        precision: parseFloat((0.0).toPrecision(precision) + '1'),
      })
      .toFixed(precision);
  }

  /**
   *  direction
   *
   * @method faker.address.direction
   * @param useAbbr return direction abbreviation. defaults to false
   */
  direction(useAbbr: boolean = false) {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr
    );
  }

  /**
   * cardinal direction
   *
   * @method faker.address.cardinalDirection
   * @param useAbbr return direction abbreviation. defaults to false
   */
  cardinalDirection(useAbbr: boolean = false): string {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction.slice(0, 4)
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr.slice(0, 4)
    );
  }

  /**
   * ordinal direction
   *
   * @method faker.address.ordinalDirection
   * @param useAbbr return direction abbreviation. defaults to false
   */
  ordinalDirection(useAbbr: boolean = false): string {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction.slice(4, 8)
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr.slice(4, 8)
    );
  }

  nearbyGPSCoordinate(
    coordinate?: number[],
    radius?: number,
    isMetric?: boolean
  ): string[] {
    function randomFloat(min: number, max: number): number {
      return Math.random() * (max - min) + min;
    }
    function degreesToRadians(degrees: number): number {
      return degrees * (Math.PI / 180.0);
    }
    function radiansToDegrees(radians: number): number {
      return radians * (180.0 / Math.PI);
    }
    function kilometersToMiles(miles: number): number {
      return miles * 0.621371;
    }
    function coordinateWithOffset(
      coordinate: number[],
      bearing: number,
      distance: number,
      isMetric: boolean
    ): number[] {
      const R = 6378.137; // Radius of the Earth (http://nssdc.gsfc.nasa.gov/planetary/factsheet/earthfact.html)
      const d = isMetric ? distance : kilometersToMiles(distance); // Distance in km

      const lat1 = degreesToRadians(coordinate[0]); //Current lat point converted to radians
      const lon1 = degreesToRadians(coordinate[1]); //Current long point converted to radians

      const lat2 = Math.asin(
        Math.sin(lat1) * Math.cos(d / R) +
          Math.cos(lat1) * Math.sin(d / R) * Math.cos(bearing)
      );

      let lon2 =
        lon1 +
        Math.atan2(
          Math.sin(bearing) * Math.sin(d / R) * Math.cos(lat1),
          Math.cos(d / R) - Math.sin(lat1) * Math.sin(lat2)
        );

      // Keep longitude in range [-180, 180]
      if (lon2 > degreesToRadians(180)) {
        lon2 = lon2 - degreesToRadians(360);
      } else if (lon2 < degreesToRadians(-180)) {
        lon2 = lon2 + degreesToRadians(360);
      }

      return [radiansToDegrees(lat2), radiansToDegrees(lon2)];
    }

    // If there is no coordinate, the best we can do is return a random GPS coordinate.
    if (coordinate === undefined) {
      return [this.faker.address.latitude(), this.faker.address.longitude()];
    }
    radius ||= 10.0;
    isMetric ||= false;

    // TODO: implement either a gaussian/uniform distribution of points in circular region.
    // Possibly include param to function that allows user to choose between distributions.

    // This approach will likely result in a higher density of points near the center.
    const randomCoord = coordinateWithOffset(
      coordinate,
      degreesToRadians(Math.random() * 360.0),
      radius,
      isMetric
    );
    return [randomCoord[0].toFixed(4), randomCoord[1].toFixed(4)];
  }

  /**
   * Return a random time zone
   *
   * @method faker.address.timeZone
   */
  timeZone(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.time_zone
    );
  }
}
