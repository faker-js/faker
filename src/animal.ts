import type { Faker } from '.';

export class Animal {
  constructor(private readonly faker: Faker) {
    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Animal.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * Returns a random dog breed.
   *
   */
  dog(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.dog);
  }

  /**
   * Returns a random cat breed.
   *
   */
  cat(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cat);
  }

  /**
   * Returns a random snake species.
   *
   */
  snake(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.snake);
  }

  /**
   * Returns a random bear species.
   *
   */
  bear(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bear);
  }

  /**
   * Returns a random lion species.
   *
   */
  lion(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.lion);
  }

  /**
   * Returns a random cetacean species.
   *
   * @method faker.animal.cetacean
   */
  cetacean(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.cetacean
    );
  }

  /**
   * Returns a random horse breed.
   *
   */
  horse(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.horse);
  }

  /**
   * Returns a random bird species.
   *
   */
  bird(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bird);
  }

  /**
   * Returns a random cow species.
   *
   */
  cow(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cow);
  }

  /**
   * Returns a random fish species.
   *
   */
  fish(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.fish);
  }

  /**
   * Returns a random crocodilian species.
   *
   */
  crocodilia(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.crocodilia
    );
  }

  /**
   * Returns a random insect species.
   *
   */
  insect(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.insect);
  }

  /**
   * Returns a random rabbit species
   *
   */
  rabbit(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.rabbit);
  }

  /**
   * Returns a random animal type.
   *
   */
  type(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.type);
  }
}
