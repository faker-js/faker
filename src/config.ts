/**
 * The possible configuration options, that can be set.
 * This type exists to be extended for plugins via type augmentation.
 *
 * The `@default` tag is used to indicate the default value, that should be used if, the config is absent.
 */
export interface FakerConfig {
  /**
   * The function used to generate the `refDate` date instance, if not provided as method param.
   * The function must return a new valid `Date` instance for every call.
   *
   * @see [Reproducible Results](https://fakerjs.dev/guide/usage.html#reproducible-results)
   * @see faker.seed(): For generating reproducible values.
   *
   * @since 9.0.0
   *
   * @default () => new Date()
   */
  defaultRefDate?: () => Date;
}
