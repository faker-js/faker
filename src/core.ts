import type { FakerConfig } from './config';
import type { LocaleDefinition } from './definitions';
import type { Randomizer } from './randomizer';

/**
 * The core grants access to the locale data, the randomizer and config settings.
 */
export interface FakerCore {
  /**
   * The locale data associated with this instance.
   *
   * Always present, but it might be empty if the locale data is not available.
   */
  readonly locale: LocaleDefinition;

  /**
   * The randomizer used to generate random values.
   */
  readonly randomizer: Randomizer;

  /**
   * The configuration settings used by this instance.
   */
  readonly config: FakerConfig;
}
