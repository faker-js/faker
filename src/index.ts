import { Faker } from './faker';
import allLocales from './locales';

export type {
  AnimalDefinitions,
  ColorDefinitions,
  CommerceDefinitions,
  CommerceProductNameDefinitions,
  CompanyDefinitions,
  DatabaseDefinitions,
  DateDefinitions,
  DateEntryDefinition,
  FinanceCurrencyEntryDefinitions,
  FinanceDefinitions,
  HackerDefinitions,
  InternetDefinitions,
  LocaleDefinition,
  /** @deprecated Use LocationDefinitions instead */
  LocationDefinitions as AddressDefinitions,
  LocationDefinitions,
  LoremDefinitions,
  MusicDefinitions,
  /** @deprecated Use PersonDefinitions instead */
  PersonDefinitions as NameDefinitions,
  PersonDefinitions,
  /** @deprecated Use PersonTitleDefinitions instead */
  PersonTitleDefinitions as NameTitleDefinitions,
  PersonTitleDefinitions,
  PhoneNumberDefinitions,
  ScienceDefinitions,
  SystemDefinitions,
  SystemMimeTypeEntryDefinitions,
  VehicleDefinitions,
  WordDefinitions,
} from './definitions';
export { FakerError } from './errors/faker-error';
export type { FakerOptions, UsableLocale, UsedLocales } from './faker';
export type { AnimalModule } from './modules/animal';
export type {
  Casing,
  ColorFormat,
  ColorModule,
  CSSFunction,
  CSSSpace,
  NumberColorFormat,
  StringColorFormat,
} from './modules/color';
export type { CommerceModule } from './modules/commerce';
export type { CompanyModule } from './modules/company';
export type { DatabaseModule } from './modules/database';
export type { DatatypeModule } from './modules/datatype';
export type { DateModule } from './modules/date';
export type { FinanceModule } from './modules/finance';
export type { GitModule } from './modules/git';
export type { HackerModule } from './modules/hacker';
export type { HelpersModule } from './modules/helpers';
export type { ImageModule } from './modules/image';
export type { InternetModule } from './modules/internet';
export type {
  /** @deprecated Use LocationModule instead */
  LocationModule as AddressModule,
  LocationModule,
} from './modules/location';
export type { LoremModule } from './modules/lorem';
export type { MusicModule } from './modules/music';
export { Sex } from './modules/person';
export type {
  /** @deprecated Use PersonModule instead */
  PersonModule as NameModule,
  PersonModule,
  SexType,
} from './modules/person';
export type { PhoneModule } from './modules/phone';
export type { RandomModule } from './modules/random';
export type { ChemicalElement, ScienceModule, Unit } from './modules/science';
export type { StringModule } from './modules/string';
export type { SystemModule } from './modules/system';
export type { VehicleModule } from './modules/vehicle';
export type { WordModule } from './modules/word';
export { Faker };

// since we are requiring the top level of faker, load all locales by default
export const faker: Faker = new Faker({
  locales: allLocales,
});
