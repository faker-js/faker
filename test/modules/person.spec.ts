import { describe, expect, it } from 'vitest';
import { faker, fakerAZ, fakerMK, fakerUK, Sex } from '../../src';
import { seededTests } from './../support/seeded-runs';
import { times } from './../support/times';

const NON_SEEDED_BASED_RUN = 5;

describe('person', () => {
  seededTests(faker, 'person', (t) => {
    t.itEach(
      'sexType',
      'gender',
      'jobTitle',
      'jobDescriptor',
      'jobArea',
      'jobType',
      'bio'
    );

    t.describeEach(
      'firstName',
      'lastName',
      'middleName',
      'prefix',
      'sex',
      'suffix'
    )((t) => t.it('noArgs').it('with sex', 'male'));

    t.describe('fullName', (t) => {
      t.it('noArgs')
        .it('with firstName', { firstName: 'John' })
        .it('with lastName', { lastName: 'Doe' })
        .it('with sex', { sex: 'female' })
        .it('with all (sex)', {
          firstName: 'John',
          lastName: 'Doe',
          sex: 'female',
        });
    });

    t.it('zodiacSign');
  });

  describe.each(times(NON_SEEDED_BASED_RUN).map(() => faker.seed()))(
    'random seeded tests for seed %i',
    () => {
      describe('firstName()', () => {
        it('should return a random first name', () => {
          const first_name = faker.person.firstName();

          expect(first_name).toBeTypeOf('string');
          expect(first_name.length).toBeGreaterThan(0);
        });

        it('should return a sex-specific first name', () => {
          let name = faker.person.firstName('female');
          expect(faker.definitions.person.female_first_name).toContain(name);

          name = faker.person.firstName('male');
          expect(faker.definitions.person.male_first_name).toContain(name);
        });

        it('should return a sex-specific first name when no sex-specific first name was defined', () => {
          const name = fakerAZ.person.firstName();
          expect([
            ...fakerAZ.definitions.person.female_first_name,
            ...fakerAZ.definitions.person.male_first_name,
          ]).toContain(name);
        });
      });

      describe('lastName()', () => {
        it('should return a random last name', () => {
          const last_name = faker.person.lastName();

          expect(last_name).toBeTypeOf('string');
          expect(last_name.length).toBeGreaterThan(0);
        });

        it('should return a sex-specific last name', () => {
          let name = fakerAZ.person.lastName('female');
          expect(fakerAZ.definitions.person.female_last_name).toContain(name);

          name = fakerAZ.person.lastName('male');
          expect(fakerAZ.definitions.person.male_last_name).toContain(name);
        });
      });

      describe('middleName()', () => {
        it('should return a random middle name', () => {
          const middle_name = faker.person.middleName();

          expect(middle_name).toBeTypeOf('string');
          expect(middle_name.length).toBeGreaterThan(0);
        });

        it('should return a middle name when passed en locale', () => {
          let name = faker.person.middleName();
          expect(faker.definitions.person.middle_name).toContain(name);

          name = faker.person.middleName('female');
          expect(faker.definitions.person.female_middle_name).toContain(name);

          name = faker.person.middleName('male');
          expect(faker.definitions.person.male_middle_name).toContain(name);
        });

        it('should return a sex-specific middle name', () => {
          let name = fakerUK.person.middleName('female');
          expect(fakerUK.definitions.person.female_middle_name).toContain(name);

          name = fakerUK.person.middleName('male');
          expect(fakerUK.definitions.person.male_middle_name).toContain(name);
        });
      });

      describe('fullName()', () => {
        it('should return a name with firstName and lastName', () => {
          const fullName = faker.person.fullName();

          expect(fullName).toBeTypeOf('string');
          expect(fullName).toContain(' ');
        });

        it('should return a female sex-specific name without firstName and lastName', () => {
          const female_specific = [
            ...(fakerMK.rawDefinitions.person?.female_prefix ?? []),
            ...(fakerMK.rawDefinitions.person?.female_first_name ?? []),
            ...(fakerMK.rawDefinitions.person?.female_last_name ?? []),
            // ...(fakerMK.rawDefinitions.person?.suffix ?? []), Not applicable
          ];

          const fullName = fakerMK.person.fullName({ sex: 'female' });

          const parts = fullName.split(' ');
          for (const part of parts) {
            expect(female_specific).toContain(part);
          }
        });

        it('should return a male sex-specific name without firstName and lastName', () => {
          const male_specific = [
            ...(fakerMK.rawDefinitions.person?.male_prefix ?? []),
            ...(fakerMK.rawDefinitions.person?.male_first_name ?? []),
            ...(fakerMK.rawDefinitions.person?.male_last_name ?? []),
            // ...(fakerMK.rawDefinitions.person?.suffix ?? []), Not applicable
          ];

          const fullName = fakerMK.person.fullName({ sex: 'male' });

          const parts = fullName.split(' ');
          for (const part of parts) {
            expect(male_specific).toContain(part);
          }
        });

        it('should return a female sex-specific name with given firstName and lastName', () => {
          const male_specific = [
            ...(fakerMK.rawDefinitions.person?.female_prefix ?? []),
            'firstName',
            'lastName',
            // ...(fakerMK.rawDefinitions.person?.suffix ?? []), Not applicable
          ];

          const fullName = fakerMK.person.fullName({
            firstName: 'firstName',
            lastName: 'lastName',
            sex: 'female',
          });

          const parts = fullName.split(' ');
          for (const part of parts) {
            expect(male_specific).toContain(part);
          }
        });

        it('should return a male sex-specific name with given firstName and lastName', () => {
          const male_specific = [
            ...(fakerMK.rawDefinitions.person?.male_prefix ?? []),
            'firstName',
            'lastName',
            // ...(fakerMK.rawDefinitions.person?.suffix ?? []), Not applicable
          ];

          const fullName = fakerMK.person.fullName({
            firstName: 'firstName',
            lastName: 'lastName',
            sex: 'male',
          });

          const parts = fullName.split(' ');
          for (const part of parts) {
            expect(male_specific).toContain(part);
          }
        });
      });

      describe('gender()', () => {
        it('should return a default gender', () => {
          const gender = faker.person.gender();

          expect(gender).toBeTypeOf('string');
          expect(faker.definitions.person.gender).toContain(gender);
        });
      });

      describe('sex()', () => {
        it('should return a sex', () => {
          const sex = faker.person.sex();

          expect(sex).toBeTypeOf('string');
          expect(faker.definitions.person.sex).toContain(sex);
        });
      });

      describe('sexType()', () => {
        it('should return a sex type', () => {
          const sexType = faker.person.sexType();

          expect(sexType).toBeTypeOf('string');
          expect(Object.values(Sex)).toContain(sexType);
        });
      });

      describe('prefix()', () => {
        it('should return a prefix', () => {
          const prefix = faker.person.prefix();

          expect(prefix).toBeTypeOf('string');
          expect(faker.definitions.person.prefix).toContain(prefix);
        });

        it('should return a female prefix with given string', () => {
          const prefix = fakerMK.person.prefix('female');

          expect(prefix).toBeTypeOf('string');
          expect(fakerMK.definitions.person.female_prefix).toContain(prefix);
        });

        it('should return a male prefix with given string', () => {
          const prefix = fakerMK.person.prefix('male');

          expect(prefix).toBeTypeOf('string');
          expect(fakerMK.definitions.person.male_prefix).toContain(prefix);
        });
      });

      describe('suffix()', () => {
        it('should return a suffix', () => {
          const suffix = faker.person.suffix();

          expect(suffix).toBeTypeOf('string');
          expect(faker.definitions.person.suffix).toContain(suffix);
        });
      });

      describe('jobTitle()', () => {
        it('should return a job title consisting of a descriptor, area, and type', () => {
          const jobTitle = faker.person.jobTitle();

          expect(jobTitle).toBeTypeOf('string');

          const [descriptor, level, job] = jobTitle.split(' ');

          expect(faker.definitions.person.title.descriptor).toContain(
            descriptor
          );
          expect(faker.definitions.person.title.level).toContain(level);
          expect(faker.definitions.person.title.job).toContain(job);
        });
      });

      describe('jobDescriptor()', () => {
        it('should return a descriptor', () => {
          const descriptor = faker.person.jobDescriptor();

          expect(descriptor).toBeTypeOf('string');

          expect(faker.definitions.person.title.descriptor).toContain(
            descriptor
          );
        });
      });

      describe('jobArea()', () => {
        it('should return a level', () => {
          const level = faker.person.jobArea();

          expect(level).toBeTypeOf('string');

          expect(faker.definitions.person.title.level).toContain(level);
        });
      });

      describe('jobType()', () => {
        it('should return a job', () => {
          const job = faker.person.jobType();

          expect(job).toBeTypeOf('string');

          expect(faker.definitions.person.title.job).toContain(job);
        });
      });

      describe('zodiacSign()', () => {
        it('returns a random zodiac sign', () => {
          const sign = faker.person.zodiacSign();

          expect(sign).toBeTypeOf('string');

          expect(faker.definitions.person.western_zodiac_sign).toContain(sign);
        });
      });
    }
  );
});
