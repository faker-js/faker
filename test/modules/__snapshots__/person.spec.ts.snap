// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`person > 42 > bio > noArgs 1`] = `"{{word.noun}} {{person.bio_supporter}}, {{person.bio_part}} {{internet.emoji}}"`;

exports[`person > 42 > bio > with options 1`] = `"{{person.bio_part}}, {{person.bio_part}}, {{internet.emoji}}"`;

exports[`person > 42 > firstName > noArgs 1`] = `"Garnet"`;

exports[`person > 42 > firstName > with sex 1`] = `"Glen"`;

exports[`person > 42 > fullName > noArgs 1`] = `"Lorene Miller"`;

exports[`person > 42 > fullName > with all (sex) 1`] = `"John Doe"`;

exports[`person > 42 > fullName > with firstName 1`] = `"John Reynolds-Miller"`;

exports[`person > 42 > fullName > with lastName 1`] = `"Lorene Doe"`;

exports[`person > 42 > fullName > with sex 1`] = `"Melanie Reynolds-Miller"`;

exports[`person > 42 > gender 1`] = `"Gender nonconforming"`;

exports[`person > 42 > jobArea 1`] = `"Identity"`;

exports[`person > 42 > jobDescriptor 1`] = `"National"`;

exports[`person > 42 > jobTitle 1`] = `"National Usability Producer"`;

exports[`person > 42 > jobType 1`] = `"Coordinator"`;

exports[`person > 42 > lastName > noArgs 1`] = `"Wiegand"`;

exports[`person > 42 > lastName > with sex 1`] = `"Wiegand"`;

exports[`person > 42 > middleName > noArgs 1`] = `"Greer"`;

exports[`person > 42 > middleName > with sex 1`] = `"Houston"`;

exports[`person > 42 > prefix > noArgs 1`] = `"Miss"`;

exports[`person > 42 > prefix > with sex 1`] = `"Mr."`;

exports[`person > 42 > sex > noArgs 1`] = `"female"`;

exports[`person > 42 > sex > with sex 1`] = `"female"`;

exports[`person > 42 > sexType 1`] = `"female"`;

exports[`person > 42 > suffix > noArgs 1`] = `"III"`;

exports[`person > 42 > suffix > with sex 1`] = `"III"`;

exports[`person > 42 > zodiacSign 1`] = `"Gemini"`;

exports[`person > 1211 > bio > noArgs 1`] = `"{{word.noun}} {{person.bio_supporter}}, {{person.bio_part}}"`;

exports[`person > 1211 > bio > with options 1`] = `"{{word.noun}} {{person.bio_supporter}}, {{person.bio_part}} {{internet.emoji}}"`;

exports[`person > 1211 > firstName > noArgs 1`] = `"Tito"`;

exports[`person > 1211 > firstName > with sex 1`] = `"Percy"`;

exports[`person > 1211 > fullName > noArgs 1`] = `"Darrel Zieme"`;

exports[`person > 1211 > fullName > with all (sex) 1`] = `"John Doe PhD"`;

exports[`person > 1211 > fullName > with firstName 1`] = `"Dr. John Fahey MD"`;

exports[`person > 1211 > fullName > with lastName 1`] = `"Darrel Doe"`;

exports[`person > 1211 > fullName > with sex 1`] = `"Miss Patti Fahey MD"`;

exports[`person > 1211 > gender 1`] = `"Trigender"`;

exports[`person > 1211 > jobArea 1`] = `"Factors"`;

exports[`person > 1211 > jobDescriptor 1`] = `"Chief"`;

exports[`person > 1211 > jobTitle 1`] = `"Chief Interactions Manager"`;

exports[`person > 1211 > jobType 1`] = `"Representative"`;

exports[`person > 1211 > lastName > noArgs 1`] = `"Trantow"`;

exports[`person > 1211 > lastName > with sex 1`] = `"Trantow"`;

exports[`person > 1211 > middleName > noArgs 1`] = `"Sawyer"`;

exports[`person > 1211 > middleName > with sex 1`] = `"Walter"`;

exports[`person > 1211 > prefix > noArgs 1`] = `"Ms."`;

exports[`person > 1211 > prefix > with sex 1`] = `"Dr."`;

exports[`person > 1211 > sex > noArgs 1`] = `"male"`;

exports[`person > 1211 > sex > with sex 1`] = `"male"`;

exports[`person > 1211 > sexType 1`] = `"male"`;

exports[`person > 1211 > suffix > noArgs 1`] = `"DVM"`;

exports[`person > 1211 > suffix > with sex 1`] = `"DVM"`;

exports[`person > 1211 > zodiacSign 1`] = `"Capricorn"`;

exports[`person > 1337 > bio > noArgs 1`] = `"{{person.bio_part}}, {{internet.emoji}}"`;

exports[`person > 1337 > bio > with options 1`] = `"{{person.bio_part}}, {{person.bio_part}}, {{internet.emoji}}"`;

exports[`person > 1337 > firstName > noArgs 1`] = `"Devyn"`;

exports[`person > 1337 > firstName > with sex 1`] = `"Ray"`;

exports[`person > 1337 > fullName > noArgs 1`] = `"Marilyn Koelpin"`;

exports[`person > 1337 > fullName > with all (sex) 1`] = `"John Doe"`;

exports[`person > 1337 > fullName > with firstName 1`] = `"John Gottlieb"`;

exports[`person > 1337 > fullName > with lastName 1`] = `"Marilyn Doe"`;

exports[`person > 1337 > fullName > with sex 1`] = `"Esther Gottlieb"`;

exports[`person > 1337 > gender 1`] = `"Demigender"`;

exports[`person > 1337 > jobArea 1`] = `"Functionality"`;

exports[`person > 1337 > jobDescriptor 1`] = `"Future"`;

exports[`person > 1337 > jobTitle 1`] = `"Future Marketing Engineer"`;

exports[`person > 1337 > jobType 1`] = `"Engineer"`;

exports[`person > 1337 > lastName > noArgs 1`] = `"Cronin"`;

exports[`person > 1337 > lastName > with sex 1`] = `"Cronin"`;

exports[`person > 1337 > middleName > noArgs 1`] = `"Dakota"`;

exports[`person > 1337 > middleName > with sex 1`] = `"Ethan"`;

exports[`person > 1337 > prefix > noArgs 1`] = `"Miss"`;

exports[`person > 1337 > prefix > with sex 1`] = `"Mr."`;

exports[`person > 1337 > sex > noArgs 1`] = `"female"`;

exports[`person > 1337 > sex > with sex 1`] = `"female"`;

exports[`person > 1337 > sexType 1`] = `"female"`;

exports[`person > 1337 > suffix > noArgs 1`] = `"I"`;

exports[`person > 1337 > suffix > with sex 1`] = `"I"`;

exports[`person > 1337 > zodiacSign 1`] = `"Taurus"`;
